modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java

```diff
@@ -628,6 +663,58 @@ public Helper(
 
                 v1 = computeCtrlPoint(x1, y1, direction, factor, n);
                 v2 = computeCtrlPoint(x2, y2, direction, -factor, n);
+
+                // Arc radius
+                r = (double) (length / properties.getFloatValue(ARC_CURVENESS));
+
+                // Arc bounding box (for Graphics2D)
+                // Formulas from https://math.stackexchange.com/questions/1781438/finding-the-center-of-a-circle-given-two-points-and-a-radius-algebraically
+                Double _xa = 0.5*(x1-x2);
+                Double _ya = 0.5*(y1-y2);
+                Double _x0 = x2+_xa;
+                Double _y0 = y2+_ya;
+                Double _a = Math.sqrt(Math.pow(_xa, 2) + Math.pow(_ya, 2));
+                Double _b = Math.sqrt(Math.pow(r, 2) - Math.pow(_a, 2));
+                Double xc = _x0 + (_b * _ya) / _a;
+                Double yc = _y0 - (_b * _xa / _a);
+                Double angle1 = Math.atan2(y1-yc, x1-xc);
+                Double angle2 = Math.atan2(y2-yc, x2-xc);
+
+                // Target radius - to start at the base of the arrow
+                final Float targetRadius = item.getData(TARGET_RADIUS);
+                // Offset due to the source node
+                if (targetRadius != null && targetRadius < 0) {
+                    Double targetOffset = this.computeTheThing(r, (double) targetRadius);
+                    angle2 += targetOffset;
+                }
+
+                // Source radius
+                final Float sourceRadius = item.getData(SOURCE_RADIUS);
+                //Avoid edge from passing the node's center:
+                if (sourceRadius != null && sourceRadius < 0) {
+                    Double sourceOffset = this.computeTheThing(r, (double) sourceRadius);
+                    angle1 -= sourceOffset;
+                }
+
+                bbx = xc-r;
+                bby = yc-r;
+                bbw = 2*r;
+                bbh = 2*r;
+                astart = -180*(angle1)/Math.PI;
+                asweep = (180*(angle1-angle2)/Math.PI+720)%360 - 360;
+            }
+
+            private Double computeTheThing(Double radius_curvature_edge, Double truncature_length) {
+                // I don't want to explain what this is. I'll try though.
```

Comment:
We want to stay away from any personal comments in the code as it's not professional